text/0005-disallow-whitespace-in-entityuid.md

```diff
@@ -0,0 +1,98 @@
+
+For example, the following syntax is valid:
+```
+permit( 
+  principal == ExampleCo :: Photoflash  ::  //This is a comment
+       :: User::"alice",
+  action,
+  resource
+);
+
+permit( 
+  principal == ExampleCo::Photoflash::User:://comment
+
+  "alice",
+  action,
+  resource
+);
+```
+
+This capability was little known, even amongst many Cedar team members, and was only identified as a result of a typo in a Cedar demo. While this behavior may be OK inside policy statements, there is a risk when policy contents such as the EntityTypeName are used as strings outside the context of policy statements, such as in the JSON schema representation. For laypersons who are not Cedar experts, these values appear to be normalized strings rather than fragments of Cedar policy syntax, leading to bugs in application code when whitespace and comments can appear in the value.
+
+Examples:
+1. The Cedar Schema format models the schema configuration under a JSON key for the namespace. Policy stores which index schemas by namespace are unlikely to recognize the need to normalize the value, leading to the possibility of storing duplicate schema definitions for "ExampleCo::Photoflash" and "ExampleCo  ::  Photoflash" and indeterminate behavior regarding which schema takes effect at runtime.
+2. Policy Stores can implement logic that relies on string comparisons against the EntityTypeName. In a real issue, an application using Cedar sought to preclude callers from passing Actions in the inline slice of entity definitions. It did so by checking if an EntityTypeName matched `.*::Action`. It presumed that `:: Action` was invalid syntax and would be rejected by the Cedar evalator, the same as any other syntatically invalid input. This resulted in a bug, as it allowed callers to bypass the extra validation that the application sought to enforce.
+3. Customers are anticipated to build meta-permissions layers that restrict callers to manipulating policy store contents for only certain namespaces. This may lead to policies such as `forbid(...) when {context.namespace = "ExampleCo::Photoflash"};`. There is a risk that an unauthorized actor could bypass this restriction by using a namespace with embedded spaces. 
+
+While it is technically possible for applications to mitigate this risk by diligently using Cedar tooling to normalize the values, the little-known nature of this Cedar behavior implies that few will know they *should* normalize the value. As a point of reference, application developers who have worked with Cedar extensively for over a year were bitten by this bug in production. Hence, this is likely to result in bugs in many other Cedar-based implementation with similar logic, risking a perception that Cedar is fragile or unsafe.
+
+## Detailed design
+In `EntityTypeName::from_str()`, whitespace would be disallowed throughout -- either before, after, or in the middle of the typename.
+
+## Drawbacks
```

Comment:
Another drawback that I don't think is called out here is the conceptual incongruity of this change. 

A namespaced type that is acceptable in a policy, e.g., `ExampleCo  ::  Photoflash  ::  User`, will be rejected when passed as a string argument to `EntityTypeName::from_str()`. This could be confusing and potentially problematic; e.g., if a schema namespace parameter is `"ExampleCo  ::  Photoflash  ::  User"` then the schema will be rejected, and thus policies will not be validated against it. 

The mitigation is to have a clear error message about what's going on. I think that should be doable because we'll have to write a custom parser for this more restricted namespaced type anyway.